File: LeetCode/14. Longest Common Prefix/solution.py

class Solution():
    """
        Author: Vishal Paudel
        Note: The try-except solution I consider elegant
        
        Accepted: true
        Runtime: 34 ms (90.29%)
        Memory Usage: 14 MB (49.39%)
    """

    def longestCommonPrefix(self, strs: List[str]) -> str:
        """
            Grab the index of the smallest one
            Search all the three words on that index
            Update the prefix
        """


        prefix = ""  # stores the current prefix
        
        try:
            # try-except for skipping searching for min word-len
            
            for index in range( len( strs[0] ) ):
            
                # this loops changes the index

                temp_prefix = ( strs[0] )[index]

                for word in strs:

                    # and this loop changes the word to viewed

                    if temp_prefix == word[index]:

                        # if the prefix has already been known

                        continue

                    else:

                        # temp_prefix does not match with letter

                        return prefix
                
                # - Update the prefix
                prefix += temp_prefix

            return prefix

        except:

            # if the index crosses the size of other words

            return prefix

# EOF

File: LeetCode/14. Longest Common Prefix/solution.cpp
class Solution {
public:
    string longestCommonPrefix(vector<string>& strs) {
        /*
         * A recursive function
         * works on only two inputs at the end
         * Recursion is beautiful! 
         */
        
        int length_strs = size(strs);
        
        if (length_strs < 1) return "";
        else if (length_strs == 1) return strs[0];
        else if (length_strs == 2) {
            
            string word1 = (strs[0]);
            string word2 = (strs[1]);

            int index_limit = min( word1.length(), word2.length() );
            string prefix = "";

            for (int index=0; index<index_limit; index++) {

                if (word1[index] == word2[index]) {

                    prefix += word1[index];

                }
                
                else {
                    
                    break;

                }

            }
            
            return prefix;

        }
        
        else {
            
            vector<string> left;
            left = vector<string> (strs.begin(), strs.end() - 1);
            
            vector<string> final = {longestCommonPrefix(left), strs[length_strs - 1]};
            
            return longestCommonPrefix(final);
        
        }
    
    }

};

// EOF
